test/clientserver_test.go

```diff
@@ -1,49 +1,246 @@
 	if err != nil {
 		t.Fatal(err)
 	}
-	infos, err := c.FindProviders(context.Background(), cid.NewCidV1(cid.Libp2pKey, h))
-	if err != nil {
-		t.Fatal(err)
+
+	ngoStart, allocStart := snapUtilizations()
+	fmt.Printf("start: goroutines=%d allocated=%d\n", ngoStart, allocStart)
+
+	timeStart := time.Now()
+
+	for i := 0; i < numIter; i++ {
+		// exercise FindProviders
+		infos, err := c.FindProviders(context.Background(), cid.NewCidV1(cid.Libp2pKey, h))
+		if err != nil {
+			t.Fatal(err)
+		}
+		if len(infos) != 1 {
+			t.Fatalf("expecting 1 result, got %d", len(infos))
+		}
+		if infos[0].ID != testAddrInfo.ID {
+			t.Errorf("expecting %#v, got %#v", testAddrInfo.ID, infos[0].ID)
+		}
+		if len(infos[0].Addrs) != 1 {
+			t.Fatalf("expecting 1 address, got %d", len(infos[0].Addrs))
+		}
+		if !infos[0].Addrs[0].Equal(testAddrInfo.Addrs[0]) {
+			t.Errorf("expecting %#v, got %#v", testAddrInfo.Addrs[0], infos[0].Addrs[0])
+		}
+		// fmt.Println(infos)
+
+		// exercise GetIPNS
+		record, err := c.GetIPNS(context.Background(), testIPNSID)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if len(record) != 1 {
+			t.Fatalf("expecting 1 result, got %d", len(record))
+		}
+		if !bytes.Equal(record[0], testIPNSRecord) {
+			t.Errorf("expecting %#v, got %#v", testIPNSRecord, record[0])
+		}
+
+		// exercise PutIPNS
+		err = c.PutIPNS(context.Background(), testIPNSID, testIPNSRecord)
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	timeEnd := time.Now()
+	avgLatency = timeEnd.Sub(timeStart) / time.Duration(numIter)
+	fmt.Printf("average roundtrip latency: %v\n", avgLatency)
+
+	ngoEnd, allocEnd := snapUtilizations()
+	fmt.Printf("end: goroutines=%d allocated=%d\n", ngoEnd, allocEnd)
+	deltaGo, deltaMem = ngoEnd-ngoStart, allocEnd-allocStart
+	fmt.Printf("diff: goroutines=%d allocated=%d\n", deltaGo, deltaMem)
+
+	return
+}
+
+type testStatistic struct {
+	total        float64
+	totalSquared float64
+	count        int64
+	max          float64
+	min          float64
+}
+
+func (s *testStatistic) Add(sample float64) {
+	s.total += sample
+	s.totalSquared += sample * sample
+	if s.count == 0 {
+		s.max = sample
+		s.min = sample
+	} else {
+		s.max = max64(s.max, sample)
+		s.min = min64(s.min, sample)
+	}
+	s.count++
+}
+
+func max64(x, y float64) float64 {
+	if x > y {
+		return x
+	}
+	return y
+}
+
+func min64(x, y float64) float64 {
+	if x < y {
+		return x
+	}
+	return y
+}
+
+func (s testStatistic) Mean() float64 {
+	return s.total / float64(s.count)
+}
+
+func (s testStatistic) Variance() float64 {
+	mean := s.Mean()
+	return s.totalSquared/float64(s.count) - mean*mean
+}
+
+func (s testStatistic) Stddev() float64 {
+	return math.Sqrt(s.Variance())
+}
+
+func (s testStatistic) MaxDeviation() float64 {
+	mean := s.Mean()
+	return max64(math.Abs(s.max-mean), math.Abs(s.min-mean))
+}
+
+func (s testStatistic) DeviatesBy(numStddev float64) bool {
+	return s.MaxDeviation()/s.Stddev() > numStddev
+}
+
+func TestClientServer(t *testing.T) {
```

Comment:
> We're pretty light on unit/correctness tests here and ofc the 4k LoC generated code isn't really human-examinable. Maybe the appropriate tests live in Edelweiss but having a few more here would be great.
> 
> Some of the edge cases that seem worth checking if not too painful include:
> 
> * How things work if I the client sends the server data not covered by the schema (i.e. does it fail, or does the base message make it through?). What about when the server responds with data not covered by the server?
This is in the spec, no? If you send more data than you need (either for client or server), things proceed successfully.
If data from the schema is missing, an error is returned. These tests belong in Edelweiss, since they are not specific to this API.

> * What about new/missing functions?
You get an error if you call a missing function. Since this is also generic, test belongs in Edelweiss. Or do you mean you want to check how the errors surface through the client/server wrappers?

> * What happens if a returned multiaddr ends in `/p2p/peerID` does the client error?
Not currently. What is the desired behavior?

> * Are our `Any` fallbacks working properly?
You mean, skipping unknown transfer protocols and unknown node types? I'll add tests.

> * Are we passing `Error` responses correctly?
I am guessing you mean if a server-side returns asynchronous error, that they show up as such on the client side. This would be tested in Edelweiss at the generated code level of client and server. If we want it all the way through to the wrapper client and server, I'll add tests in this repo.

It sounds like adding tests on the wrapper client/server will make everyone feel good. With a lot of these tests there are two levels of possible testing: generate client to generated server, or wrapped client to wrapped server. The former should be in Edelweiss and the latter here (obviously).

Reply:
> What happens if a returned multiaddr ends in /p2p/peerID does the client error?
>> Not currently. What is the desired behavior?

[The spec](https://github.com/ipfs/specs/blob/6bdb7b2751038e1d0212a40494ea8fd4018f384c/REFRAME.md?plain=1#L168) indicates we shouldn't end in `/p2p/peerID` so our options seem to be:
1. Be very strict and return an error if any part of the response is non-compliant
2. Be a little strict and drop multiaddrs with `/p2p/peerID` at the end perhaps with some log output informing about the error
3. Be loose and rewrite the multiaddr to not end with `/p2p/peerID` assuming the extra peerID matches the peerID in the `ID` field. If it doesn't match then log.

2 seems reasonable to me WDYT?

Reply:
Yes, I think 2 is the way to go.